File: agents/scala/fw-agent/src/main/scala/afl/agent/model/TaskDocument.scala
package fw.agent.model

import org.mongodb.scala.bson.{BsonNull, BsonValue, Document}
import org.mongodb.scala.documentToUntypedDocument

import scala.jdk.CollectionConverters.*

/** Represents a task document from the AFL tasks collection. */
case class TaskDocument(
    uuid: String,
    name: String,
    runnerId: String,
    workflowId: String,
    flowId: String,
    stepId: String,
    state: String,
    created: Long,
    updated: Long,
    error: Option[Document],
    taskListName: String,
    dataType: String,
    data: Option[Document]
)

object TaskDocument:
  /** Deserialize a MongoDB document into a TaskDocument. */
  def fromBson(doc: Document): TaskDocument =
    val underlying: org.bson.Document = doc
    TaskDocument(
      uuid = underlying.getString("uuid"),
      name = underlying.getString("name"),
      runnerId = underlying.getString("runner_id"),
      workflowId = underlying.getString("workflow_id"),
      flowId = underlying.getString("flow_id"),
      stepId = underlying.getString("step_id"),
      state = underlying.getString("state"),
      created = underlying.getLong("created"),
      updated = underlying.getLong("updated"),
      error = optionalSubDoc(doc, "error"),
      taskListName = underlying.getString("task_list_name"),
      dataType = underlying.getString("data_type"),
      data = optionalSubDoc(doc, "data")
    )

  private def optionalSubDoc(doc: Document, key: String): Option[Document] =
    doc.get[BsonValue](key) match
      case Some(bv) if bv.isDocument => Some(Document(bv.asDocument()))
      case _                         => None

  extension (task: TaskDocument)
    /** Serialize to a MongoDB document. */
    def toBson: Document =
      Document(
        "uuid" -> task.uuid,
        "name" -> task.name,
        "runner_id" -> task.runnerId,
        "workflow_id" -> task.workflowId,
        "flow_id" -> task.flowId,
        "step_id" -> task.stepId,
        "state" -> task.state,
        "created" -> task.created,
        "updated" -> task.updated,
        "error" -> task.error.map(_.toBsonDocument).getOrElse(BsonNull()),
        "task_list_name" -> task.taskListName,
        "data_type" -> task.dataType,
        "data" -> task.data.map(_.toBsonDocument).getOrElse(BsonNull())
      )

File: agents/java/fw-agent/src/main/java/afl/agent/model/TaskDocument.java
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package afl.agent.model;

import org.bson.Document;

import java.util.Map;

/**
 * Represents a task in the tasks collection.
 */
public record TaskDocument(
        String uuid,
        String name,
        String runnerId,
        String workflowId,
        String flowId,
        String stepId,
        String state,
        long created,
        long updated,
        String errorMessage,
        String taskListName,
        String dataType,
        Map<String, Object> data
) {

    /**
     * Creates a TaskDocument from a MongoDB document.
     */
    public static TaskDocument fromDocument(Document doc) {
        String errorMessage = null;
        Document error = doc.get("error", Document.class);
        if (error != null) {
            errorMessage = error.getString("message");
        }

        @SuppressWarnings("unchecked")
        Map<String, Object> data = doc.get("data", Map.class);

        return new TaskDocument(
                doc.getString("uuid"),
                doc.getString("name"),
                doc.getString("runner_id"),
                doc.getString("workflow_id"),
                doc.getString("flow_id"),
                doc.getString("step_id"),
                doc.getString("state"),
                doc.getLong("created"),
                doc.getLong("updated"),
                errorMessage,
                doc.getString("task_list_name"),
                doc.getString("data_type"),
                data
        );
    }

    /**
     * Converts to a MongoDB document.
     */
    public Document toDocument() {
        Document doc = new Document()
                .append("uuid", uuid)
                .append("name", name)
                .append("runner_id", runnerId)
                .append("workflow_id", workflowId)
                .append("flow_id", flowId)
                .append("step_id", stepId)
                .append("state", state)
                .append("created", created)
                .append("updated", updated)
                .append("task_list_name", taskListName);

        if (dataType != null) {
            doc.append("data_type", dataType);
        }
        if (data != null) {
            doc.append("data", new Document(data));
        }
        if (errorMessage != null) {
            doc.append("error", new Document("message", errorMessage));
        }

        return doc;
    }
}
